src/Init/Data/List/Lemmas.lean

```diff
@@ -707,33 +741,105 @@ theorem get_map (f : α → β) {l n} :
+theorem map_congr_left (h : ∀ a ∈ l, f a = g a) : map f l = map g l :=
+  map_inj_left.2 h
 
-theorem exists_of_mem_map (h : b ∈ map f l) : ∃ a, a ∈ l ∧ f a = b := mem_map.1 h
+theorem map_inj : map f = map g ↔ f = g := by
+  constructor
+  · intro h; ext a; replace h := congrFun h [a]; simpa using h
+  · intro h; subst h; rfl
+
+@[simp] theorem map_eq_nil {f : α → β} {l : List α} : map f l = [] ↔ l = [] := by
+  constructor <;> exact fun _ => match l with | [] => rfl
+
+theorem eq_nil_of_map_eq_nil {f : α → β} {l : List α} (h : map f l = []) : l = [] :=
+  map_eq_nil.mp h
+
+theorem map_eq_cons {f : α → β} {l : List α} :
+    map f l = b :: l₂ ↔ l.head?.map f = some b ∧ l.tail?.map (map f) = some l₂ := by
+  induction l <;> simp_all
+
+theorem map_eq_cons' {f : α → β} {l : List α} :
+    map f l = b :: l₂ ↔ ∃ a l₁, l = a :: l₁ ∧ f a = b ∧ map f l₁ = l₂ := by
+  cases l
+  case nil => simp
+  case cons a l₁ =>
+    simp only [map_cons, cons.injEq]
+    constructor
+    · rintro ⟨rfl, rfl⟩
+      exact ⟨a, l₁, ⟨rfl, rfl⟩, ⟨rfl, rfl⟩⟩
+    · rintro ⟨a, l₁, ⟨rfl, rfl⟩, ⟨rfl, rfl⟩⟩
+      constructor <;> rfl
+
+theorem map_eq_foldr (f : α → β) (l : List α) : map f l = foldr (fun a bs => f a :: bs) [] l := by
+  induction l <;> simp [*]
+
+@[simp] theorem set_map {f : α → β} {l : List α} {n : Nat} {a : α} :
+    (map f l).set n (f a) = map f (l.set n a) := by
+  induction l generalizing n with
+  | nil => simp
+  | cons b l ih => cases n <;> simp_all
+
+@[simp] theorem head_map (f : α → β) (l : List α) (w) :
+    head (map f l) w = f (head l (by simpa using w)) := by
+  cases l
+  · simp at w
+  · simp_all
+
+@[simp] theorem head?_map (f : α → β) (l : List α) : head? (map f l) = (head? l).map f := by
+  cases l <;> rfl
+
+@[simp] theorem headD_map (f : α → β) (l : List α) (a : α) : headD (map f l) (f a) = f (headD l a) := by
+  cases l <;> rfl
+
+@[simp] theorem tail?_map (f : α → β) (l : List α) : tail? (map f l) = (tail? l).map (map f) := by
+  cases l <;> rfl
+
+@[simp] theorem tailD_map (f : α → β) (l : List α) (l' : List α) :
+    tailD (map f l) (map f l') = map f (tailD l l') := by
+  cases l <;> rfl
+
+@[simp] theorem getLast_map (f : α → β) (l : List α) (h) :
+    getLast (map f l) h = f (getLast l (by simpa using h)) := by
+  cases l
+  · simp at h
+  · simp only [← getElem_cons_length _ _ _ rfl]
+    simp only [map_cons]
+    simp only [← getElem_cons_length _ _ _ rfl]
+    simp only [← map_cons, getElem_map]
+    simp
+
+@[simp] theorem getLast?_map (f : α → β) (l : List α) : getLast? (map f l) = (getLast? l).map f := by
+  cases l
+  · simp
+  · rw [getLast?_eq_getLast, getLast?_eq_getLast, getLast_map] <;> simp
+
+@[simp] theorem getLastD_map (f : α → β) (l : List α) (a : α) : getLastD (map f l) (f a) = f (getLastD l a) := by
+  cases l
+  · simp
+  · rw [getLastD_eq_getLast?, getLastD_eq_getLast?, getLast?_map] <;> simp
+
+@[simp] theorem map_append (f : α → β) : ∀ l₁ l₂, map f (l₁ ++ l₂) = map f l₁ ++ map f l₂ := by
```

Comment:
This duplicates `f`. I still think this lemma should be simp, but maybe we can give a rationale in the design section above that duplicating higher order argument (mapped functions, predicates) is ok, because they are often small?